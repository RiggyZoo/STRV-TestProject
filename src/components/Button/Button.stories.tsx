import React, { ReactNode } from 'react'
import { Story } from '@storybook/react'
import { Button, Buttons, ButtonSize } from './Button'

interface IButton {
  children: string | ReactNode
  name: string
  theme: Buttons
  size: ButtonSize
  loading: boolean
}

export default {
  title: 'Components/Button',

  component: Button,
}

const Template: Story<IButton> = (args) => <Button {...args} />

export const Default = Template.bind({})

Default.args = {
  children: 'Button',
  theme: Buttons.red,
  size: ButtonSize.main,
  loading: true,
}
