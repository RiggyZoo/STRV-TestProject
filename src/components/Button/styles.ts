import styled from 'styled-components'
import Cursor from '../../assets/icons/Pointer.png'
import { colors } from '../../styles/themes'
import { IButtonProps } from './Button'

export const DefaultButton = styled.button<IButtonProps>`
  font-size: ${({ size }) => (size === 'small' ? '0.875rem' : '1rem')};
  position: relative;
  min-width: ${({ size }) => (size === 'small' ? '6.25rem' : '15rem')};
  padding: ${({ size }) =>
    size === 'small' ? '0.625rem 0 0.5rem 0' : '0.8125rem 0 0.75rem 0'};
  line-height: ${({ size }) => (size === 'small' ? '0.875rem' : '2rem')};
  cursor: url(${Cursor}), pointer;
  text-transform: uppercase;
  border-radius: 4px;
  background: ${colors.primary};
  text-align: center;
  border: none;
  letter-spacing: 1px;
  color: ${colors.white};
  font-family: Hind-Regular, sans-serif;
  transition: all 0.2s;

  & > span {
    background-color: ${({ loading }) => loading && 'red'};
    visibility: ${({ loading }) => loading && 'hidden'};
  }
  &:hover {
    background: ${colors.primaryHover};
  }
`
export const RedButton = styled(DefaultButton)`
  background: ${colors.secondary};
  &:hover {
    background: ${colors.secondaryHover};
  }
`
export const GreyButton = styled(DefaultButton)`
  background: ${colors.grey};
  &:hover {
    background: ${colors.greyHover};
  }
`

export const RefreshButton = styled(DefaultButton)`
  background: ${colors.inputValue};
  &:hover {
    background: ${colors.inputValue};
  }
`
