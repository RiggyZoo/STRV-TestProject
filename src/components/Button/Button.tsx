import React, { FC } from 'react'
import { StyledComponent } from 'styled-components'
import { Loader } from '../Loader'
import { DefaultButton, RedButton, GreyButton, RefreshButton } from './styles'

export interface IButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  theme: 'green' | 'red' | 'grey' | 'refresh'
  size: 'main' | 'small'
  loading?: any
  ref?: any
}

const Button: FC<IButtonProps> = ({
  children,
  theme = 'green',
  loading,
  ref,
  size = 'small',
  ...rest
}) => {
  const themes = {
    green: DefaultButton,
    red: RedButton,
    grey: GreyButton,
    refresh: RefreshButton,
  }

  const Style: StyledComponent<'button', any, IButtonProps, any> =
    themes[theme || 'default']

  return (
    <Style ref={ref} type="button" size={size} $loading={loading} {...rest}>
      {loading && (
        <Loader
          top={size === 'main' ? '12%' : '10%'}
          right="50%"
          size={size === 'main' ? 'normal' : 'small'}
        />
      )}{' '}
      <span>{children}</span>
    </Style>
  )
}

export { Button }
