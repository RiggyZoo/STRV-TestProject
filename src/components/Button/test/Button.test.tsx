import { Button } from '../Button'
import * as ReactDOM from 'react-dom'
import React from 'react'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'

describe('Button component tests', () => {
  let container: HTMLDivElement

  beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
    ReactDOM.render(<Button theme="red" size="small" />, container)
  })

  afterEach(() => {
    document.body.removeChild(container)
    container.remove()
  })

  it('Renders without crashing ', function () {
    ReactDOM.render(<Button theme="red" size="small" />, container)
  })
  it('Renders correctly', function () {
    const { getByTestId } = render(
      <Button theme={'red'} size={'main'}>
        Click
      </Button>,
      {
        container: container,
      },
    )
    expect(getByTestId('button')).toHaveTextContent('Click')
  })

  it('Matches to snapshot ', function () {
    const tree = renderer
      .create(
        <Button theme={'red'} size={'main'}>
          Click
        </Button>,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
