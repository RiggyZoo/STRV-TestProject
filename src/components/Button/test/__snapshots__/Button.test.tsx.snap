// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests Matches to snapshot  1`] = `
<button
  className="sc-dkPtRN sc-hKwDye gNicrT culDtV"
  data-testid="button"
  size="main"
  type="button"
>
   
  <span>
    Click
  </span>
</button>
`;
